// @proto-kit/library@0.1.1-develop.833/dist/protocol/VanillaProtocolModules.js
import { AccountStateHook, BlockHeightHook, BlockProver, StateTransitionProver, LastStateRootBlockHook, } from "@proto-kit/protocol";
import { PrivateKey } from "o1js";
import { TransactionFeeHook } from "../hooks/TransactionFeeHook";
export class VanillaProtocolModules {
    static mandatoryModules(additionalModules) {
        return {
            StateTransitionProver,
            BlockProver,
            AccountState: AccountStateHook,
            BlockHeight: BlockHeightHook,
            LastStateRoot: LastStateRootBlockHook,
            ...additionalModules,
        };
    }
    static with(additionalModules) {
        return {
            ...VanillaProtocolModules.mandatoryModules(additionalModules),
            TransactionFee: TransactionFeeHook,
        };
    }
    static mandatoryConfig() {
        return {
            BlockProver: {},
            StateTransitionProver: {},
            AccountState: {},
            BlockHeight: {},
            LastStateRoot: {},
        };
    }
    static defaultConfig() {
        return {
            ...VanillaProtocolModules.mandatoryConfig(),
            TransactionFee: {
                tokenId: 0n,
                feeRecipient: PrivateKey.random().toPublicKey().toBase58(),
                baseFee: 0n,
                perWeightUnitFee: 0n,
                methods: {},
            },
        };
    }
}
