// @proto-kit/sequencer@0.1.1-develop.833/dist/protocol/production/sequencing/BlockProducerModule.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
import { inject } from "tsyringe";
import { log, noop } from "@proto-kit/common";
import { ACTIONS_EMPTY_HASH } from "@proto-kit/protocol";
import { MethodIdResolver, MethodParameterEncoder, Runtime, } from "@proto-kit/module";
import { Provable } from "o1js";
import { sequencerModule, SequencerModule, } from "../../../sequencer/builder/SequencerModule";
import { BlockWithResult } from "../../../storage/model/Block";
import { CachedStateService } from "../../../state/state/CachedStateService";
import { TransactionExecutionService } from "./TransactionExecutionService";
export let BlockProducerModule = class BlockProducerModule extends SequencerModule {
    constructor(mempool, messageStorage, unprovenStateService, unprovenMerkleStore, blockQueue, blockTreeStore, executionService, methodIdResolver, runtime) {
        super();
        this.mempool = mempool;
        this.messageStorage = messageStorage;
        this.unprovenStateService = unprovenStateService;
        this.unprovenMerkleStore = unprovenMerkleStore;
        this.blockQueue = blockQueue;
        this.blockTreeStore = blockTreeStore;
        this.executionService = executionService;
        this.methodIdResolver = methodIdResolver;
        this.runtime = runtime;
        this.productionInProgress = false;
    }
    allowEmptyBlock() {
        return this.config.allowEmptyBlock ?? true;
    }
    prettyPrintBlockContents(block) {
        block.transactions.forEach((tx, i) => {
            const methodName = this.methodIdResolver.getMethodNameFromId(tx.tx.methodId.toBigInt());
            if (!methodName)
                return;
            const module = this.runtime.resolve(methodName[0]);
            const paramEncoder = MethodParameterEncoder.fromMethod(module, methodName[1]);
            log.info("---------------------------------------");
            log.info(`Transaction #${i}`);
            log.info("Sender:", tx.tx.sender.toBase58(), "Nonce:", tx.tx.nonce.toBigInt());
            log.info(`Method: ${methodName?.join(".")}`);
            log.info();
            if (log.getLevel() <= log.levels.INFO) {
                Provable.log("Arguments:", paramEncoder.decode(tx.tx.argsFields, tx.tx.auxiliaryData));
            }
            log.info(`Status: ${tx.status.toBoolean()}`, tx.statusMessage !== undefined ? `Reason: ${tx.statusMessage}` : "");
        });
        if (block.transactions.length > 0) {
            log.info("---------------------------------------");
        }
    }
    async tryProduceBlock() {
        if (!this.productionInProgress) {
            try {
                const block = await this.produceBlock();
                if (block === undefined) {
                    if (!this.allowEmptyBlock()) {
                        log.info("No transactions in mempool, skipping production");
                    }
                    else {
                        log.error("Something wrong happened, skipping block");
                    }
                    return undefined;
                }
                log.info(`Produced block (${block.transactions.length} txs)`);
                this.prettyPrintBlockContents(block);
                // Generate metadata for next block
                // TODO: make async of production in the future
                const result = await this.executionService.generateMetadataForNextBlock(block, this.unprovenMerkleStore, this.blockTreeStore, true);
                return {
                    block,
                    result,
                };
            }
            catch (error) {
                if (error instanceof Error) {
                    throw error;
                }
                else {
                    log.error(error);
                }
            }
            finally {
                this.productionInProgress = false;
            }
        }
        return undefined;
    }
    async collectProductionData() {
        const txs = await this.mempool.getTxs();
        const parentBlock = await this.blockQueue.getLatestBlock();
        if (parentBlock === undefined) {
            log.debug("No block metadata given, assuming first block, generating genesis metadata");
        }
        const messages = await this.messageStorage.getMessages(parentBlock?.block.toMessagesHash.toString() ??
            ACTIONS_EMPTY_HASH.toString());
        const metadata = parentBlock ?? BlockWithResult.createEmpty();
        log.debug(`Block collected, ${txs.length} txs, ${messages.length} messages`);
        return {
            txs: messages.concat(txs),
            metadata,
        };
    }
    async produceBlock() {
        this.productionInProgress = true;
        const { txs, metadata } = await this.collectProductionData();
        // Skip production if no transactions are available for now
        if (txs.length === 0 && !this.allowEmptyBlock()) {
            return undefined;
        }
        const cachedStateService = new CachedStateService(this.unprovenStateService);
        const block = await this.executionService.createBlock(cachedStateService, txs, metadata, this.allowEmptyBlock());
        await cachedStateService.mergeIntoParent();
        this.productionInProgress = false;
        return block;
    }
    async start() {
        noop();
    }
};
BlockProducerModule = __decorate([
    sequencerModule(),
    __param(0, inject("Mempool")),
    __param(1, inject("MessageStorage")),
    __param(2, inject("UnprovenStateService")),
    __param(3, inject("UnprovenMerkleStore")),
    __param(4, inject("BlockQueue")),
    __param(5, inject("BlockTreeStore")),
    __param(7, inject("MethodIdResolver")),
    __param(8, inject("Runtime")),
    __metadata("design:paramtypes", [Object, Object, Object, Object, Object, Object, TransactionExecutionService,
        MethodIdResolver,
        Runtime])
], BlockProducerModule);
