// @trpc/react-query@11.0.0-rc.332/dist/shared/proxy/utilsProxy.js
'use strict';

var client = require('@trpc/client');
var unstableCoreDoNotImport = require('@trpc/server/unstable-core-do-not-import');
var context = require('../../internals/context.js');
var getQueryKey = require('../../internals/getQueryKey.js');

const getQueryType = (utilName)=>{
    switch(utilName){
        case 'fetch':
        case 'ensureData':
        case 'prefetch':
        case 'getData':
        case 'setData':
            return 'query';
        case 'fetchInfinite':
        case 'prefetchInfinite':
        case 'getInfiniteData':
        case 'setInfiniteData':
            return 'infinite';
        case 'cancel':
        case 'invalidate':
        case 'refetch':
        case 'reset':
            return 'any';
    }
};
/**
 * @internal
 */ function createRecursiveUtilsProxy(context, key) {
    return unstableCoreDoNotImport.createRecursiveProxy((opts)=>{
        const path = [
            key,
            ...opts.path
        ];
        const utilName = path.pop();
        const args = [
            ...opts.args
        ];
        const input = args.shift(); // args can now be spread when input removed
        const queryType = getQueryType(utilName);
        const queryKey = getQueryKey.getQueryKeyInternal(path, input, queryType);
        const contextMap = {
            fetch: ()=>context.fetchQuery(queryKey, ...args),
            fetchInfinite: ()=>context.fetchInfiniteQuery(queryKey, args[0]),
            prefetch: ()=>context.prefetchQuery(queryKey, ...args),
            prefetchInfinite: ()=>context.prefetchInfiniteQuery(queryKey, args[0]),
            ensureData: ()=>context.ensureQueryData(queryKey, ...args),
            invalidate: ()=>context.invalidateQueries(queryKey, ...args),
            reset: ()=>context.resetQueries(queryKey, ...args),
            refetch: ()=>context.refetchQueries(queryKey, ...args),
            cancel: ()=>context.cancelQuery(queryKey, ...args),
            setData: ()=>{
                context.setQueryData(queryKey, args[0], args[1]);
            },
            setInfiniteData: ()=>{
                context.setInfiniteQueryData(queryKey, args[0], args[1]);
            },
            getData: ()=>context.getQueryData(queryKey),
            getInfiniteData: ()=>context.getInfiniteQueryData(queryKey)
        };
        return contextMap[utilName]();
    });
}
/**
 * @internal
 */ function createReactQueryUtils(context$1) {
    return unstableCoreDoNotImport.createFlatProxy((key)=>{
        const contextName = key;
        if (contextName === 'client') {
            return client.createTRPCClientProxy(context$1.client);
        }
        if (context.contextProps.includes(contextName)) {
            return context$1[contextName];
        }
        return createRecursiveUtilsProxy(context$1, key);
    });
}
/**
 * @internal
 */ function createQueryUtilsProxy(context) {
    return unstableCoreDoNotImport.createFlatProxy((key)=>{
        return createRecursiveUtilsProxy(context, key);
    });
}

exports.createQueryUtilsProxy = createQueryUtilsProxy;
exports.createReactQueryUtils = createReactQueryUtils;
exports.getQueryType = getQueryType;
