// @proto-kit/library@0.1.1-develop.833/dist/sequencer/MonoSequencerModules.js
import { BlockProducerModule, LocalTaskWorkerModule, PrivateMempool, UnprovenProducerModule, VanillaTaskWorkerModules, } from "@proto-kit/sequencer";
import { GraphqlSequencerModule, GraphqlServer, DefaultGraphqlModules, } from "@proto-kit/api";
export class MonoSequencerModules {
    static worker(queue, additionalModules) {
        return {
            LocalTaskWorkerModule: LocalTaskWorkerModule.from(VanillaTaskWorkerModules.withoutSettlement()),
            TaskQueue: queue,
            ...additionalModules,
        };
    }
    static with(queue, database, baselayer, trigger, additionalModules) {
        return {
            Database: database,
            Mempool: PrivateMempool,
            BaseLayer: baselayer,
            BlockProducerModule: BlockProducerModule,
            UnprovenProducerModule: UnprovenProducerModule,
            BlockTrigger: trigger,
            TaskQueue: queue,
            GraphqlServer: GraphqlServer,
            Graphql: GraphqlSequencerModule.from({
                modules: DefaultGraphqlModules.with({}),
            }),
            ...additionalModules,
        };
    }
    static defaultConfig() {
        return {
            UnprovenProducerModule: {},
            GraphqlServer: {
                port: 8080,
                host: "0.0.0.0",
                graphiql: true,
            },
            Graphql: {
                QueryGraphqlModule: {},
                MempoolResolver: {},
                BlockStorageResolver: {},
                NodeStatusResolver: {},
                MerkleWitnessResolver: {},
                UnprovenBlockResolver: {},
            },
            Mempool: {},
            BlockProducerModule: {},
        };
    }
}
