// @trpc/server@11.0.0-rc.332/dist/unstable-core-do-not-import.js
'use strict';

var createProxy = require('./unstable-core-do-not-import/createProxy.js');
var formatter = require('./unstable-core-do-not-import/error/formatter.js');
var getErrorShape = require('./unstable-core-do-not-import/error/getErrorShape.js');
var TRPCError = require('./unstable-core-do-not-import/error/TRPCError.js');
var getHTTPStatusCode = require('./unstable-core-do-not-import/http/getHTTPStatusCode.js');
var resolveHTTPResponse = require('./unstable-core-do-not-import/http/resolveHTTPResponse.js');
var batchStreamFormatter = require('./unstable-core-do-not-import/http/batchStreamFormatter.js');
var contentType = require('./unstable-core-do-not-import/http/contentType.js');
var toURL = require('./unstable-core-do-not-import/http/toURL.js');
var initTRPC = require('./unstable-core-do-not-import/initTRPC.js');
var middleware = require('./unstable-core-do-not-import/middleware.js');
var parser = require('./unstable-core-do-not-import/parser.js');
var procedure = require('./unstable-core-do-not-import/procedure.js');
var procedureBuilder = require('./unstable-core-do-not-import/procedureBuilder.js');
var rootConfig = require('./unstable-core-do-not-import/rootConfig.js');
var router = require('./unstable-core-do-not-import/router.js');
var codes = require('./unstable-core-do-not-import/rpc/codes.js');
var parseTRPCMessage = require('./unstable-core-do-not-import/rpc/parseTRPCMessage.js');
var transformer = require('./unstable-core-do-not-import/transformer.js');
var types = require('./unstable-core-do-not-import/types.js');
var utils = require('./unstable-core-do-not-import/utils.js');



exports.createFlatProxy = createProxy.createFlatProxy;
exports.createRecursiveProxy = createProxy.createRecursiveProxy;
exports.defaultFormatter = formatter.defaultFormatter;
exports.getErrorShape = getErrorShape.getErrorShape;
exports.TRPCError = TRPCError.TRPCError;
exports.getCauseFromUnknown = TRPCError.getCauseFromUnknown;
exports.getTRPCErrorFromUnknown = TRPCError.getTRPCErrorFromUnknown;
exports.getHTTPStatusCode = getHTTPStatusCode.getHTTPStatusCode;
exports.getHTTPStatusCodeFromError = getHTTPStatusCode.getHTTPStatusCodeFromError;
exports.resolveHTTPResponse = resolveHTTPResponse.resolveHTTPResponse;
exports.getBatchStreamFormatter = batchStreamFormatter.getBatchStreamFormatter;
exports.getJsonContentTypeInputs = contentType.getJsonContentTypeInputs;
exports.toURL = toURL.toURL;
exports.initTRPC = initTRPC.initTRPC;
exports.createInputMiddleware = middleware.createInputMiddleware;
exports.createMiddlewareFactory = middleware.createMiddlewareFactory;
exports.createOutputMiddleware = middleware.createOutputMiddleware;
exports.experimental_standaloneMiddleware = middleware.experimental_standaloneMiddleware;
exports.middlewareMarker = middleware.middlewareMarker;
exports.getParseFn = parser.getParseFn;
exports.procedureTypes = procedure.procedureTypes;
exports.createBuilder = procedureBuilder.createBuilder;
exports.unsetMarker = procedureBuilder.unsetMarker;
exports.isServerDefault = rootConfig.isServerDefault;
exports.callProcedure = router.callProcedure;
exports.createCallerFactory = router.createCallerFactory;
exports.createRouterFactory = router.createRouterFactory;
exports.mergeRouters = router.mergeRouters;
exports.TRPC_ERROR_CODES_BY_KEY = codes.TRPC_ERROR_CODES_BY_KEY;
exports.TRPC_ERROR_CODES_BY_NUMBER = codes.TRPC_ERROR_CODES_BY_NUMBER;
exports.parseTRPCMessage = parseTRPCMessage.parseTRPCMessage;
exports.defaultTransformer = transformer.defaultTransformer;
exports.getDataTransformer = transformer.getDataTransformer;
exports.transformResult = transformer.transformResult;
exports.transformTRPCResponse = transformer.transformTRPCResponse;
exports.ERROR_SYMBOL = types.ERROR_SYMBOL;
exports.isFunction = utils.isFunction;
exports.isObject = utils.isObject;
exports.mergeWithoutOverrides = utils.mergeWithoutOverrides;
exports.omitPrototype = utils.omitPrototype;
