// @proto-kit/api@0.1.1-develop.833/dist/graphql/VanillaGraphqlModules.js
import { MempoolResolver } from "./modules/MempoolResolver";
import { QueryGraphqlModule } from "./modules/QueryGraphqlModule";
import { BatchStorageResolver } from "./modules/BatchStorageResolver";
import { NodeStatusResolver } from "./modules/NodeStatusResolver";
import { BlockResolver } from "./modules/BlockResolver";
import { MerkleWitnessResolver } from "./modules/MerkleWitnessResolver";
export class VanillaGraphqlModules {
    static with(additionalModules) {
        return {
            MempoolResolver,
            QueryGraphqlModule,
            BatchStorageResolver,
            NodeStatusResolver,
            BlockResolver,
            MerkleWitnessResolver,
            ...additionalModules,
        };
    }
    static defaultConfig() {
        return {
            MempoolResolver: {},
            QueryGraphqlModule: {},
            BatchStorageResolver: {},
            NodeStatusResolver: {},
            BlockResolver: {},
            MerkleWitnessResolver: {},
        };
    }
}
