// @trpc/server@11.0.0-rc.332/dist/adapters/aws-lambda/index.mjs
import { TRPCError } from '../../unstable-core-do-not-import/error/TRPCError.mjs';
import { resolveHTTPResponse } from '../../unstable-core-do-not-import/http/resolveHTTPResponse.mjs';
import '../../unstable-core-do-not-import/rootConfig.mjs';
import { getPath, getHTTPMethod, isPayloadV1, transformHeaders, isPayloadV2, UNKNOWN_PAYLOAD_FORMAT_VERSION_ERROR_MESSAGE } from './utils.mjs';

/**
 * If you're making an adapter for tRPC and looking at this file for reference, you should import types and functions from `@trpc/server` and `@trpc/server/http`
 *
 * @example
 * ```ts
 * import type { AnyTRPCRouter } from '@trpc/server'
 * import type { HTTPBaseHandlerOptions } from '@trpc/server/http'
 * ```
 */ // @trpc/server
function lambdaEventToHTTPRequest(event) {
    const query = new URLSearchParams();
    for (const [key, value] of Object.entries(event.queryStringParameters ?? {})){
        if (typeof value !== 'undefined') {
            query.append(key, value);
        }
    }
    let body;
    if (event.body && event.isBase64Encoded) {
        body = Buffer.from(event.body, 'base64').toString('utf8');
    } else {
        body = event.body;
    }
    return {
        method: getHTTPMethod(event),
        query: query,
        headers: event.headers,
        body: body
    };
}
function tRPCOutputToAPIGatewayOutput(event, response) {
    if (isPayloadV1(event)) {
        const resp = {
            statusCode: response.status,
            body: response.body ?? '',
            headers: transformHeaders(response.headers ?? {})
        };
        return resp;
    } else if (isPayloadV2(event)) {
        const resp1 = {
            statusCode: response.status,
            body: response.body ?? undefined,
            headers: transformHeaders(response.headers ?? {})
        };
        return resp1;
    } else {
        throw new TRPCError({
            code: 'INTERNAL_SERVER_ERROR',
            message: UNKNOWN_PAYLOAD_FORMAT_VERSION_ERROR_MESSAGE
        });
    }
}
function awsLambdaRequestHandler(opts) {
    return async (event, context)=>{
        const req = lambdaEventToHTTPRequest(event);
        const path = getPath(event);
        const createContext = async (innerOpts)=>{
            return await opts.createContext?.({
                event,
                context,
                ...innerOpts
            });
        };
        const response = await resolveHTTPResponse({
            ...opts,
            createContext,
            req,
            path,
            error: null,
            onError (o) {
                opts?.onError?.({
                    ...o,
                    req: event
                });
            }
        });
        return tRPCOutputToAPIGatewayOutput(event, response);
    };
}

export { UNKNOWN_PAYLOAD_FORMAT_VERSION_ERROR_MESSAGE, awsLambdaRequestHandler, getHTTPMethod, getPath, isPayloadV1, isPayloadV2, transformHeaders };
