// @proto-kit/library@0.1.1-develop.833/dist/types/UInt224.js
import { UIntX } from "./UInt";
import { Field, UInt32, UInt64 } from "o1js";
import { UInt112 } from "./UInt112";
export class UInt224 extends UIntX {
    /**
     * Static method to create a {@link UIntX} with value `0`.
     */
    static get zero() {
        return UInt224.from(Field(0));
    }
    /**
     * Static method to create a {@link UIntX} with value `1`.
     */
    static get one() {
        return UInt224.from(Field(1));
    }
    static MAXINT() {
        return new UInt224(UIntX.maxIntField(UInt224.NUM_BITS));
    }
    static check(x) {
        const actual = x.value.rangeCheckHelper(UInt224.NUM_BITS);
        UIntX.assertionFunction(actual.equals(x.value));
    }
    static from(x) {
        if (x instanceof UInt64 || x instanceof UInt32 || x instanceof UIntX) {
            x = x.value;
        }
        return new UInt224(UInt224.checkConstant(Field(x), UInt224.NUM_BITS));
    }
    constructor(value) {
        super(value, UInt224.NUM_BITS, {
            creator: (x) => new UInt224(x),
            from: (x) => UInt224.from(x),
        });
    }
    toUInt112() {
        const uint112 = new UInt112(this.value);
        UInt112.check(uint112);
        return uint112;
    }
}
// eslint-disable-next-line @typescript-eslint/no-magic-numbers
UInt224.NUM_BITS = 224;
