// @proto-kit/library@0.1.1-develop.833/dist/types/UInt32.js
import { Field, UInt32 as o1UInt32 } from "o1js";
import { UIntX } from "./UInt";
export class UInt32 extends UIntX {
    /**
     * Static method to create a {@link UIntX} with value `0`.
     */
    static get zero() {
        return UInt32.from(Field(0));
    }
    /**
     * Static method to create a {@link UIntX} with value `1`.
     */
    static get one() {
        return UInt32.from(Field(1));
    }
    static MAXINT() {
        return new UInt32(UIntX.maxIntField(UInt32.NUM_BITS));
    }
    static check(x) {
        const actual = x.value.rangeCheckHelper(UInt32.NUM_BITS);
        UIntX.assertionFunction(actual.equals(x.value));
    }
    static from(x) {
        if (x instanceof UInt32 || x instanceof o1UInt32) {
            x = x.value;
        }
        return new UInt32(UInt32.checkConstant(Field(x), UInt32.NUM_BITS));
    }
    constructor(value) {
        super(value, UInt32.NUM_BITS, {
            creator: (x) => new UInt32(x),
            from: (x) => UInt32.from(x),
        });
    }
}
// eslint-disable-next-line @typescript-eslint/no-magic-numbers
UInt32.NUM_BITS = 32;
