// @proto-kit/sdk@0.1.1-develop.833/dist/query/NetworkStateQuery.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { injectable } from "tsyringe";
import { NetworkState } from "@proto-kit/protocol";
import { UInt64 } from "o1js";
let NetworkStateQuery = class NetworkStateQuery {
    constructor(blockService) {
        this.blockService = blockService;
    }
    get currentNetworkState() {
        return this.getCurrentNetworkState();
    }
    async getCurrentNetworkState() {
        const height = await this.blockService.getCurrentBlockHeight();
        return new NetworkState({
            block: {
                height: UInt64.from(height),
            },
        });
    }
};
NetworkStateQuery = __decorate([
    injectable(),
    __metadata("design:paramtypes", [Object])
], NetworkStateQuery);
export { NetworkStateQuery };
