// next@15.0.0-canary.148/dist/client/components/react-dev-overlay/internal/container/StaticIndicator.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "StaticIndicator", {
    enumerable: true,
    get: function() {
        return StaticIndicator;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _Toast = require("../components/Toast");
const _LightningBolt = require("../icons/LightningBolt");
const _CloseIcon = require("../icons/CloseIcon");
function StaticIndicator(param) {
    let { dispatcher } = param;
    return /*#__PURE__*/ (0, _jsxruntime.jsxs)(_Toast.Toast, {
        className: "nextjs-static-indicator-toast-wrapper",
        children: [
            /*#__PURE__*/ (0, _jsxruntime.jsx)(_LightningBolt.LightningBolt, {}),
            /*#__PURE__*/ (0, _jsxruntime.jsx)("span", {
                children: "Static route "
            }),
            /*#__PURE__*/ (0, _jsxruntime.jsx)("button", {
                onClick: ()=>{
                    dispatcher == null ? void 0 : dispatcher.onStaticIndicator(false);
                },
                className: "nextjs-toast-hide-button",
                style: {
                    marginLeft: 'var(--size-gap)'
                },
                children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_CloseIcon.CloseIcon, {})
            })
        ]
    });
}

if ((typeof exports.default === 'function' || (typeof exports.default === 'object' && exports.default !== null)) && typeof exports.default.__esModule === 'undefined') {
  Object.defineProperty(exports.default, '__esModule', { value: true });
  Object.assign(exports.default, exports);
  module.exports = exports.default;
}
