// @proto-kit/sequencer@0.1.1-develop.833/dist/worker/queue/BullQueue.js
import { MetricsTime, Queue, QueueEvents, Worker } from "bullmq";
import { log } from "@proto-kit/common";
/**
 * TaskQueue implementation for BullMQ
 */
export class BullQueue {
    constructor(redis, options = {}) {
        this.redis = redis;
        this.options = options;
    }
    createWorker(name, executor, options) {
        const worker = new Worker(name, async (job) => JSON.stringify(await executor(job.data)), {
            concurrency: options?.concurrency ?? 1,
            connection: this.redis,
            metrics: { maxDataPoints: MetricsTime.ONE_HOUR * 24 },
        });
        // We have to do this, because we want to prevent the worker from crashing
        worker.on("error", (error) => {
            log.error("Worker threw error:");
            log.error(error);
        });
        return {
            async close() {
                await worker.close();
            },
        };
    }
    async getQueue(queueName) {
        const queue = new Queue(queueName, {
            connection: this.redis,
        });
        const events = new QueueEvents(queueName, { connection: this.redis });
        await queue.drain();
        const { options } = this;
        return {
            name: queueName,
            async addTask(payload) {
                const job = await queue.add(queueName, payload, {
                    attempts: options.retryAttempts ?? 2,
                });
                return { taskId: job.id };
            },
            async onCompleted(listener) {
                events.on("completed", async (result) => {
                    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
                    await listener(JSON.parse(result.returnvalue));
                });
                await events.waitUntilReady();
            },
            async close() {
                await events.close();
                await queue.close();
            },
        };
    }
}
